In [1]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

In [2]:
bookings = pd.read_csv('data/bookings.csv', encoding = 'windows-1251', sep = ';')

In [3]:
bookings

Hotel  Is Canceled  Lead Time arrival full date  \
0       Resort Hotel            0        342        2015-07-01   
1       Resort Hotel            0        737        2015-07-01   
2       Resort Hotel            0          7        2015-07-01   
3       Resort Hotel            0         13        2015-07-01   
4       Resort Hotel            0         14        2015-07-01   
...              ...          ...        ...               ...   
119385    City Hotel            0         23        2017-08-30   
119386    City Hotel            0        102        2017-08-31   
119387    City Hotel            0         34        2017-08-31   
119388    City Hotel            0        109        2017-08-31   
119389    City Hotel            0        205        2017-08-29   

        Arrival Date Year Arrival Date Month  Arrival Date Week Number  \
0                    2015               July                        27   
1                    2015               July                        27   
2                    2015               July                        27   
3                    2015               July                        27   
4                    2015               July                        27   
...                   ...                ...                       ...   
119385               2017             August                        35   
119386               2017             August                        35   
119387               2017             August                        35   
119388               2017             August                        35   
119389               2017             August                        35   

        Arrival Date Day of Month  Stays in Weekend nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        Stays in week nights  ...  Adults  Children  Babies  Meal Country  \
0                          0  ...       2       0.0       0    BB     PRT   
1                          0  ...       2       0.0       0    BB     PRT   
2                          1  ...       1       0.0       0    BB     GBR   
3                          1  ...       1       0.0       0    BB     GBR   
4                          2  ...       2       0.0       0    BB     GBR   
...                      ...  ...     ...       ...     ...   ...     ...   
119385                     5  ...       2       0.0       0    BB     BEL   
119386                     5  ...       3       0.0       0    BB     FRA   
119387                     5  ...       2       0.0       0    BB     DEU   
119388                     5  ...       2       0.0       0    BB     GBR   
119389                     7  ...       2       0.0       0    HB     DEU   

       Reserved Room Type Assigned room type customer type Reservation Status  \
0                       C                  C     Transient          Check-Out   
1                       C                  C     Transient          Check-Out   
2                       A                  C     Transient          Check-Out   
3                       A                  A     Transient          Check-Out   
4                       A                  A     Transient          Check-Out   
...                   ...                ...           ...                ...   
119385                  A                  A     Transient          Check-Out   
119386         

In [4]:
bookings_head = bookings.head(7) # вывести первые 7 строчек таблицы
bookings_head

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  

[7 rows x 21 columns]

In [5]:
bookings.shape 

(119390, 21)

In [6]:
bookings.dtypes #проверка типов данных в колонках

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

In [7]:
def adsc(name):
    new_name=name.lower().replace(' ', '_') #Создал функцию, которая меняет большие буквы на маленькие 
    return new_name                         # и делает замену пробелов на _

In [8]:
bookings = bookings.rename(columns=adsc) 
bookings

hotel  is_canceled  lead_time arrival_full_date  \
0       Resort Hotel            0        342        2015-07-01   
1       Resort Hotel            0        737        2015-07-01   
2       Resort Hotel            0          7        2015-07-01   
3       Resort Hotel            0         13        2015-07-01   
4       Resort Hotel            0         14        2015-07-01   
...              ...          ...        ...               ...   
119385    City Hotel            0         23        2017-08-30   
119386    City Hotel            0        102        2017-08-31   
119387    City Hotel            0         34        2017-08-31   
119388    City Hotel            0        109        2017-08-31   
119389    City Hotel            0        205        2017-08-29   

        arrival_date_year arrival_date_month  arrival_date_week_number  \
0                    2015               July                        27   
1                    2015               July                        27   
2                    2015               July                        27   
3                    2015               July                        27   
4                    2015               July                        27   
...                   ...                ...                       ...   
119385               2017             August                        35   
119386               2017             August                        35   
119387               2017             August                        35   
119388               2017             August                        35   
119389               2017             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  ...  adults  children  babies  meal country  \
0                          0  ...       2       0.0       0    BB     PRT   
1                          0  ...       2       0.0       0    BB     PRT   
2                          1  ...       1       0.0       0    BB     GBR   
3                          1  ...       1       0.0       0    BB     GBR   
4                          2  ...       2       0.0       0    BB     GBR   
...                      ...  ...     ...       ...     ...   ...     ...   
119385                     5  ...       2       0.0       0    BB     BEL   
119386                     5  ...       3       0.0       0    BB     FRA   
119387                     5  ...       2       0.0       0    BB     DEU   
119388                     5  ...       2       0.0       0    BB     GBR   
119389                     7  ...       2       0.0       0    HB     DEU   

       reserved_room_type assigned_room_type customer_type reservation_status  \
0                       C                  C     Transient          Check-Out   
1                       C                  C     Transient          Check-Out   
2                       A                  C     Transient          Check-Out   
3                       A                  A     Transient          Check-Out   
4                       A                  A     Transient          Check-Out   
...                   ...                ...           ...                ...   
119385                  A                  A     Transient          Check-Out   
119386         

In [9]:
bookings.query("reservation_status == 'Check-Out'").groupby('country') \
        .agg({'reservation_status':'count'}) \
        .sort_values('reservation_status', ascending=False)

reservation_status
country                    
PRT                   21071
GBR                    9676
FRA                    8481
ESP                    6391
DEU                    6069
...                     ...
SMR                       1
MLI                       1
SLE                       1
MAC                       1
PLW                       1

[165 rows x 1 columns]

In [10]:
p=bookings.groupby('hotel') \
          .agg({'stays_total_nights':'mean'}) \
          .sort_values('stays_total_nights', ascending=False) 
p.round(2) 

stays_total_nights
hotel                           
Resort Hotel                4.32
City Hotel                  2.98

In [11]:
bookings.query('assigned_room_type != reserved_room_type') 

hotel  is_canceled  lead_time arrival_full_date  \
2       Resort Hotel            0          7        2015-07-01   
12      Resort Hotel            0         68        2015-07-01   
15      Resort Hotel            0         68        2015-07-01   
17      Resort Hotel            0         12        2015-07-01   
18      Resort Hotel            0          0        2015-07-01   
...              ...          ...        ...               ...   
119273    City Hotel            0        213        2017-08-28   
119274    City Hotel            0        213        2017-08-28   
119289    City Hotel            0         25        2017-08-30   
119297    City Hotel            0        332        2017-08-31   
119357    City Hotel            0         47        2017-08-31   

        arrival_date_year arrival_date_month  arrival_date_week_number  \
2                    2015               July                        27   
12                   2015               July                        27   
15                   2015               July                        27   
17                   2015               July                        27   
18                   2015               July                        27   
...                   ...                ...                       ...   
119273               2017             August                        35   
119274               2017             August                        35   
119289               2017             August                        35   
119297               2017             August                        35   
119357               2017             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
2                               1                        0   
12                              1                        0   
15                              1                        0   
17                              1                        0   
18                              1                        0   
...                           ...                      ...   
119273                         28                        1   
119274                         28                        1   
119289                         30                        0   
119297                         31                        0   
119357                         31                        1   

        stays_in_week_nights  ...  adults  children  babies  meal country  \
2                          1  ...       1       0.0       0    BB     GBR   
12                         4  ...       2       0.0       0    BB     USA   
15                         4  ...       2       0.0       0    BB     IRL   
17                         1  ...       2       0.0       0    BB     IRL   
18                         1  ...       2       0.0       0    BB     FRA   
...                      ...  ...     ...       ...     ...   ...     ...   
119273                     3  ...       1       0.0       0    HB     PRT   
119274                     3  ...       1       0.0       0    HB     PRT   
119289                     3  ...       3       0.0       0    BB     ITA   
119297                     2  ...       2       0.0       0    BB     GBR   
119357                     3  ...       1       0.0       0    SC     PRT   

       reserved_room_type assigned_room_type    customer_type  \
2                       A                  C        Transient   
12                      D                  E        Transient   
15                      D                  E        Transient   
17                      A                  E        Transient   
18                      A                  G        Transient   
...                   ...                ...              ...   
119273                  A                  K  Transient-Party   
119274                  A                  K  Transient-Party   
119289                  E                  F        Transient   
119297       

In [12]:
#На какой месяц чаще всего оформляли бронь в 2016 году? Изменился ли самый популярный месяц в 2017?
bookings.groupby(['arrival_date_month', 'arrival_date_year']) \
        .agg({'reservation_status_date' : 'count'}) \
        .sort_values('arrival_date_month', ascending=False)

reservation_status_date
arrival_date_month arrival_date_year                         
September          2016                                  5394
                   2015                                  5114
October            2016                                  6203
                   2015                                  4957
November           2016                                  4454
                   2015                                  2340
May                2017                                  6313
                   2016                                  5478
March              2017                                  4970
                   2016                                  4824
June               2016                                  5292
                   2017                                  5647
July               2017                                  5313
                   2016                                  4572
                   2015                                  2776
January            2017                                  3681
                   2016                                  2248
February           2017                                  4177
                   2016                                  3891
December           2016                                  3860
                   2015                                  2920
August             2017                                  4925
                   2016                                  5063
                   2015                                  3889
April              2017                                  5661
                   2016                                  5428

In [13]:
#Сгруппируйте данные по годам, а затем проверьте, на какой месяц 
#бронирования отеля типа City Hotel отменялись чаще всего в 2015? 2016? 2017? 

bookings.query("is_canceled == 1 and hotel == 'City Hotel'") \
        .groupby(['arrival_date_month', 'arrival_date_year'], as_index=False) \
        .agg({'reservation_status' : 'count'}) \
        .sort_values('reservation_status', ascending=False)

arrival_date_month  arrival_date_year  reservation_status
19                May               2017                2217
23            October               2016                1947
1               April               2017                1926
15               June               2017                1808
14               June               2016                1720
25          September               2016                1567
24          September               2015                1543
0               April               2016                1539
18                May               2016                1436
21           November               2016                1360
13               July               2017                1324
22            October               2015                1321
17              March               2017                1278
3              August               2016                1247
2              August               2015                1232
4              August               2017                1123
16              March               2016                1108
6            December               2016                1072
10            January               2017                1044
12               July               2016                1043
8            February               2017                 971
11               July               2015                 939
7            February               2016                 930
5            December               2015                 668
9             January               2016                 438
20           November               2015                 301

In [14]:
#Посмотрите на числовые характеристики трёх колонок: adults, children и babies.
#Какая из них имеет наибольшее среднее значение?

bookings[{'adults', 'children', 'babies'}].describe()

adults         babies       children
count  119390.000000  119390.000000  119386.000000
mean        1.856403       0.007949       0.103890
std         0.579261       0.097436       0.398561
min         0.000000       0.000000       0.000000
25%         2.000000       0.000000       0.000000
50%         2.000000       0.000000       0.000000
75%         2.000000       0.000000       0.000000
max        55.000000      10.000000      10.000000

**Ответ :** abults с средним значением 1.856403

In [15]:
#Создайте колонку total_kids, объединив столбцы children и babies. 
#Для отелей какого типа среднее значение переменной оказалось наибольшим?
#В качестве ответа укажите наибольшее среднее total_kids, округлив до 2 знаков после точки.

bookings['total_kids'] = bookings.children + bookings.babies

In [16]:
bookings.groupby('hotel', as_index=False) \
        .agg({'total_kids' : 'mean'}) \
        .sort_values('total_kids', ascending=False) \
        .round(2)

hotel  total_kids
1  Resort Hotel        0.14
0    City Hotel        0.10

**Ответ :** наибольшее среднее значение переменной total_kids равно 0.14 для Resort Hotel